numpy/core/src/multiarray/buffer.c

```diff
@@ -596,6 +630,13 @@ _buffer_get_info(PyObject *obj)
             item = PyList_GetItem(item_list, PyList_GET_SIZE(item_list) - 1);
             old_info = (_buffer_info_t*)PyLong_AsVoidPtr(item);
 
+            /*
+             * TODO: In rare cases when an array is both C- and
+             *       F-contiguous, and buffers are exported alternating
+             *       we may always have a cache miss even though the correct
+             *       info had already been exported before (causing a memory
+             *       leak).
+             */
```

Comment:
This seems very easy to fix - just make `key` the tuple `PyTuple_Pack(2, PyLong_FromVoidPtr((void*)obj), PyBool_FromLong(f_contiguous))` (with error checking).

Reply:
Hmmm, it does mean you have to delete two elements on every single array deallocation, though.

Reply:
So it trades of something that should never happen, with a small speed penalty on every deallocation right now.

Reply:
Sorry, I guess that is not true, it might happen, but its unlikely common. The other alternative is to always probe 2 buffers into the past for "reuse", and actually, I guess only when the array is both C- and F-contiguous, which is quick to check.

Reply:
If you think using a `(pointer, order)` key is two expensive, then probably best to add that justification to a source comment.

Reply:
I think the issue was that I had neither of those ideas when writing that comment. And I do hate 98% solutions. I will have a quick look whether inspecting two list elements is easy, if not, I will expand the comment.

Reply:
OK, I pushed a fix, too often just staring at it and wondering takes more time than the fix :(. I have added a test, which trips if run with `pytest-leaks`. I have confirmed that it fails before and passes after.